src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+        private static int s_objectTypeCount;
+
+        internal readonly int _objectID = Interlocked.Increment(ref s_objectTypeCount);
+
+        internal int ObjectID => _objectID;
+
+        public IHandler<PreLoginHandlerContext> NextHandler { get; set; }
+
+        public async ValueTask Handle(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            await PreloginPacketHandler.CreatePreLoginAndSend(context, isAsync, ct).ConfigureAwait(false);
+            if (context.HasError)
+            {
+                return;
+            }
+            await ReadPreLoginresponse(context, isAsync, ct).ConfigureAwait(false);
+
+            if (context.HasError)
+            {
+                return;
+            }
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Consume the Prelogin response from the server.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async Task ReadPreLoginresponse(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            context.ConnectionContext.MarsCapable = context.ConnectionContext.ConnectionString.MARS; // Assign default value
+            context.ConnectionContext.FedAuthRequired = false;
+
+            TdsStream tdsStream = context.ConnectionContext.TdsStream;
+
+            // Look into the first byte to see if we are connecting to a 6.5 or earlier server.
+            int option = await tdsStream.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+
+            if (option == 0xaa)
+            {
+                // If the first byte is 0xAA, we are connecting to a 6.5 or earlier server, which
+                // is not supported.
+                throw SQL.InvalidSQLServerVersionUnknown();
+            }
+
+            byte[] payload = new byte[tdsStream.PacketDataLeft];
+
+            int payloadOffset = 0;
+            int offset = 0;
+
+            while (option != (byte)PreLoginOptions.LASTOPT)
+            {
+                switch (option)
+                {
+                    case (int)PreLoginOptions.VERSION:
+
+                        payloadOffset = payload[offset++] << 8 | payload[offset++];
+                        offset += PAYLOAD_LENGTH_SIZE_IN_BYTES; // Skip the payload length
+
+                        byte majorVersion = payload[payloadOffset];
+
+                        bool is2005OrLater = majorVersion >= 9;
+                        if (!is2005OrLater)
+                        {
+                            context.ConnectionContext.MarsCapable = false;  // If pre-2005, MARS not supported.
+                        }
+
+                        break;
+
+                    case (int)PreLoginOptions.ENCRYPT:
```

Comment:
Please make sure Encrypt case runs regardless of ENCRYPT option received in prelogin response, see [commit](https://github.com/dotnet/SqlClient/commit/7cb40938b85cb3901f2c0e30c63da63ee531de1f) for ref.

Reply:
@cheenamalhotra  I dont understand. Isn't this code already doing what the linked commit intended?

Reply:
I don't think so.? Consider the case where server doesn't send ENCRYPT flag in PreLoginOptions, would you still set InternalEncryptionOption based on user's request "Encrypt=true"?

Reply:
If the server doesn't sent encrypt flag, (applicable to TDS74 only, since for 8, we always make sure to do TLS in the beginning and stay in TLS), but if the client had requested it, then the `InternalEncryptionOption` is ON. Check the prelogin send part.

IF the client didnt request encryption which means `InternalEncryptionOption` is OFF, and the Server also says that it doesn't support encryption, then we set `InternalEncryptionOption` to LOGIN to make sure that login is encrypted. This is the case where the server and client both negotiate encryption in TDS74. As a result, the client enforces LOGIN negotiation.

After this check the Tds74Handler where we check if the client has requested Encryption (It is always true because if atleast default to LOGIN encryption), and then if the server has said that it doesnt support encryption, then we throw. Tds74Handler again.

Reply:
Resolving. Please reopen if the explanation is unsatisfactory.